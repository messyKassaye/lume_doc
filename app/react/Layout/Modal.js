import PropTypes from 'prop-types';
import React, { Component } from 'react';
import ReactModal from 'react-modal';

export default class Modal extends Component {
  render() {
    const style = { overlay: { zIndex: 100, backgroundColor: 'rgba(0, 0, 0, 0.75)' } };
    const type = this.props.type || 'success';
    return (
      <ReactModal
        style={style}
        className={`modal-dialog modal-${type}`}
        isOpen={this.props.isOpen}
        contentLabel=""
        ariaHideApp={false}
      >
        <div className="modal-content">{this.props.children}</div>
      </ReactModal>
    );
  }
}

const childrenType = PropTypes.oneOfType([PropTypes.object, PropTypes.array, PropTypes.string]);

Modal.propTypes = {
  isOpen: PropTypes.bool,
  type: PropTypes.string,
  children: childrenType,
};

const Body = ({ children }) => <div className="modal-body">{children}</div>;
Body.propTypes = { children: childrenType };

const Header = ({ children }) => <div className="modal-header">{children}</div>;
Header.propTypes = { children: childrenType };

const Footer = ({ children }) => <div className="modal-footer">{children}</div>;
Footer.propTypes = { children: childrenType };

const Title = ({ children }) => <h4 className="modal-title">{children}</h4>;
Title.propTypes = { children: childrenType };

const Close = ({ onClick }) => (
  <button type="button" className="close" onClick={onClick}>
    <span aria-hidden="true">&times;</span>
    <span className="sr-only">Close</span>
  </button>
);
Close.propTypes = { onClick: PropTypes.func };

Modal.Body = Body;
Modal.Header = Header;
Modal.Footer = Footer;
Modal.Title = Title;
Modal.Close = Close;
