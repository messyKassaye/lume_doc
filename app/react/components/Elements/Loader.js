import React, { Component } from 'react';
import './scss/loader.scss';

class Loader extends Component {
  render() {
    return (
      <div className="cs-loader">
        <div className="cs-loader-inner">
          <label> ●</label>
          <label> ●</label>
          <label> ●</label>
        </div>
      </div>
    );
  }
}

export default Loader;
