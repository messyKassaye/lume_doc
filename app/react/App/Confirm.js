import PropTypes from 'prop-types';
import React, { Component } from 'react';
import { t } from 'app/I18N';

import Modal from 'app/Layout/Modal';
import Loader from 'app/components/Elements/Loader';

export class Confirm extends Component {
  static getDerivedStateFromProps(newProps, state) {
    if (newProps.accept !== state.accept) {
      return { isOpen: true, accept: newProps.accept };
    }

    return null;
  }

  constructor(props) {
    super(props);
    this.state = {
      isOpen: props.isOpen,
      isLoading: props.isLoading,
      confirmInputValue: '',
    };

    this.accept = this.accept.bind(this);
    this.cancel = this.cancel.bind(this);
    this.close = this.close.bind(this);
    this.handleInput = this.handleInput.bind(this);
  }

  close() {
    this.setState({ isOpen: false, confirmInputValue: '', isLoading: false });
  }

  accept() {
    if (this.props.accept) {
      const actionResponse = this.props.accept();
      if (actionResponse && actionResponse instanceof Promise) {
        this.setState({ isLoading: true });
        actionResponse.then(this.close);
        actionResponse.catch(this.close);
        return;
      }
    }
    this.close();
  }

  cancel() {
    if (this.props.cancel) {
      this.props.cancel();
    }
    this.close();
  }

  handleInput(e) {
    this.setState({ confirmInputValue: e.target.value });
  }

  renderExtraConfirm() {
    return (
      <>
        <p> If you want to continue, please type &#39;{this.props.extraConfirmWord}&#39; </p>
        <input type="text" onChange={this.handleInput} value={this.state.confirmInputValue} />
      </>
    );
  }

  render() {
    const { type } = this.props;
    return (
      <Modal isOpen={this.state.isOpen} type={type}>
        <Modal.Body>
          <h4>{this.props.title}</h4>
          <p>{this.props.message}</p>
          {this.props.extraConfirm && !this.state.isLoading && this.renderExtraConfirm()}
          {this.state.isLoading && <Loader />}
        </Modal.Body>

        {!this.state.isLoading && (
          <Modal.Footer>
            {!this.props.noCancel && (
              <button type="button" className="btn btn-default cancel-button" onClick={this.cancel}>
                {t('System', 'Cancel')}
              </button>
            )}
            <button
              type="button"
              disabled={
                this.props.extraConfirm &&
                this.state.confirmInputValue !== this.props.extraConfirmWord
              }
              className={`btn confirm-button btn-${type}`}
              onClick={this.accept}
            >
              {t('System', 'Accept')}
            </button>
          </Modal.Footer>
        )}
      </Modal>
    );
  }
}

Confirm.defaultProps = {
  isLoading: false,
  extraConfirm: false,
  isOpen: false,
  noCancel: false,
  type: 'danger',
  title: 'Confirm action',
  message: 'Are you sure you want to continue?',
  extraConfirmWord: 'CONFIRM',
};

Confirm.propTypes = {
  isLoading: PropTypes.bool,
  extraConfirm: PropTypes.bool,
  extraConfirmWord: PropTypes.string,
  isOpen: PropTypes.bool,
  noCancel: PropTypes.bool,
  accept: PropTypes.func,
  cancel: PropTypes.func,
  type: PropTypes.string,
  title: PropTypes.string,
  message: PropTypes.string,
};

export default Confirm;
