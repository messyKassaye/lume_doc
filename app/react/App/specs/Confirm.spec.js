import React from 'react';
import { shallow } from 'enzyme';

import Modal from 'app/Layout/Modal';
import Loader from 'app/components/Elements/Loader';
import Confirm from '../Confirm';

describe('CantDeleteTemplateAlert', () => {
  let component;
  let props;
  let instance;

  beforeEach(() => {
    props = {
      cancel: jasmine.createSpy('cancel'),
      isOpen: false,
    };
  });

  const render = () => {
    component = shallow(<Confirm {...props} />);
    instance = component.instance();
  };

  it('should render a default closed modal', () => {
    render();
    expect(component.find(Modal).props().isOpen).toBe(false);
  });

  it('noCancel option should hide the cancel button', () => {
    props.noCancel = true;
    render();
    expect(component).toMatchSnapshot();
  });

  it('extraConfirm option should render a confirm input', () => {
    props.extraConfirm = true;
    render();
    expect(component).toMatchSnapshot();
  });

  describe('when clicking ok button', () => {
    it('should call accept function', () => {
      props.isOpen = true;
      props.accept = jasmine.createSpy('accept');
      render();
      component.find('.btn-danger').simulate('click');
      expect(props.accept).toHaveBeenCalled();
      expect(instance.state.isOpen).toBe(false);
      expect(instance.state.isLoading).toBe(false);
    });

    describe('when the action is async', () => {
      let resolve;
      let reject;
      let promise;
      beforeEach(() => {
        promise = new Promise((_resolve, _reject) => {
          resolve = _resolve;
          reject = _reject;
        }).catch(() => {});
        props.accept = jasmine.createSpy('accept');
        props.accept.and.returnValue(promise);
        props.isOpen = true;
        render();
      });

      it('should show a loading state until the promise resolves and render a Loader', done => {
        component.find('.btn-danger').simulate('click');
        expect(props.accept).toHaveBeenCalled();
        expect(instance.state.isOpen).toBe(true);
        expect(instance.state.isLoading).toBe(true);
        expect(component.find(Loader).length).toBe(1);
        resolve();
        promise.then(() => {
          expect(instance.state.isOpen).toBe(false);
          expect(instance.state.isLoading).toBe(false);
          done();
        });
      });

      it('should show a loading state until the promise rejects', done => {
        component.find('.btn-danger').simulate('click');
        expect(props.accept).toHaveBeenCalled();
        expect(instance.state.isOpen).toBe(true);
        expect(instance.state.isLoading).toBe(true);
        expect(component.find(Loader).length).toBe(1);
        reject();
        promise.then(() => {
          expect(instance.state.isOpen).toBe(false);
          expect(instance.state.isLoading).toBe(false);
          done();
        });
      });
    });
  });

  describe('when clicking cancel button', () => {
    it('should call cancel function', () => {
      render();
      component.find('.btn-default').simulate('click');
      expect(props.cancel).toHaveBeenCalled();
    });
  });
});
