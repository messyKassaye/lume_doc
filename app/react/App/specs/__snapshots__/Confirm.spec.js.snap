// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CantDeleteTemplateAlert extraConfirm option should render a confirm input 1`] = `
<Modal
  isOpen={false}
  type="danger"
>
  <Body>
    <h4>
      Confirm action
    </h4>
    <p>
      Are you sure you want to continue?
    </p>
    <p>
       If you want to continue, please type '
      CONFIRM
      ' 
    </p>
    <input
      onChange={[Function]}
      type="text"
      value=""
    />
  </Body>
  <Footer>
    <button
      className="btn btn-default cancel-button"
      onClick={[Function]}
      type="button"
    >
      Cancel
    </button>
    <button
      className="btn confirm-button btn-danger"
      disabled={true}
      onClick={[Function]}
      type="button"
    >
      Accept
    </button>
  </Footer>
</Modal>
`;

exports[`CantDeleteTemplateAlert noCancel option should hide the cancel button 1`] = `
<Modal
  isOpen={false}
  type="danger"
>
  <Body>
    <h4>
      Confirm action
    </h4>
    <p>
      Are you sure you want to continue?
    </p>
  </Body>
  <Footer>
    <button
      className="btn confirm-button btn-danger"
      disabled={false}
      onClick={[Function]}
      type="button"
    >
      Accept
    </button>
  </Footer>
</Modal>
`;
